### ISSUE TITLE
Compilation with Docker not working

### ISSUE THREAD

**Issue by pabluk** [miekg/coredns#6303]
**What happened**:

As described on the [README.md](https://github.com/coredns/coredns#compilation-with-docker) file I tried to build coredns using Docker but the command:
```shell
docker run --rm -i -t -v $PWD:/v -w /v golang:1.21 make
```
failed with this error:
```
...
fatal: detected dubious ownership in repository at '/v'
To add an exception for this directory, call:

	git config --global --add safe.directory /v
error obtaining VCS status: exit status 128
	Use -buildvcs=false to disable VCS stamping.
make: *** [Makefile:16: coredns] Error 1
```

**What you expected to happen**:

I expected to have a `coredns` binary on the current working directory.

**How to reproduce it (as minimally and precisely as possible)**:
```shell
# clone master branch of this repo (latest commit: 1a57825f0e16bdeb9e4d509165681f3c9647537e)
git clone git@github.com:coredns/coredns.git
cd coredns
docker run --rm -i -t -v $PWD:/v -w /v golang:1.21 make
```

**Anything else we need to know?**:

I know this isn't a bug on coredns itself, it's just documentation bug related to git and docker but it can be very useful for newcomers.

The error seems related to the user inside the container (`root`) and the ownership of the files under `$PWD` because the user running the `git` command must be the owner of `.git` and the current working directories. As I can see here:
```
$ docker run --rm -i -t -v $PWD:/v -w /v golang:1.21
root@672b53987beb:/v# ls -al
total 240
drwxrwxr-x 13 1000 1000  4096 Sep  5 19:34 .
drwxr-xr-x  1 root root  4096 Sep  5 20:14 ..
...
drwxrwxr-x  8 1000 1000  4096 Sep  5 20:14 .git
...
root@672b53987beb:/v# git status
fatal: detected dubious ownership in repository at '/v'
To add an exception for this directory, call:

	git config --global --add safe.directory /v
root@672b53987beb:/v# 
```
I've also tried to change the user running inside the container using `-u=$UID` but the build cache directory is by default under `/.cache` owned by `root` so it fail too:
```
docker run --rm -i -t -u=$UID -v $PWD:/v -w /v golang:1.21 make
CGO_ENABLED=0  go build -v -ldflags="-s -w -X github.com/coredns/coredns/coremain.GitCommit=1a57825f" -o coredns
failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied
make: *** [Makefile:16: coredns] Error 1

```
**Workaround**:
Due to the git error `fatal: detected dubious ownership in repository at '/v'` the only workaround I found, _and it's ugly_, is to change the ownership of `.` and `.git` directories: 
```shell
git clone git@github.com:coredns/coredns.git
cd coredns
sudo chown root:root .
sudo chown root:root .git
docker run --rm -i -t -v $PWD:/v -w /v golang:1.21 make
```

**Environment**:
- the version of CoreDNS:
  CoreDNS-1.11.1
  linux/amd64, go1.21.0, 1a57825f

- Corefile: N/A

- OS (e.g: `cat /etc/os-release`): 
  $ lsb_release -a
  Distributor ID:	Ubuntu
  Description:	Ubuntu 22.04.1 LTS
  Release:	22.04
  Codename:	jammy

- Others:
  $ docker --version
  Docker version 24.0.5, build ced0996

**Comment by chrisohaver** [comment]
I am able to build using the documented docker command without error.
My `.` and `.git` directories are not owned by root.

**Comment by chrisohaver** [comment]
Your user is displayed as a numeric ID, which can be caused by the user being deleted.    Perhaps git is sensitive to this, hence the "dubious owner" error.

**Comment by pabluk** [comment]
Hi @chrisohaver and thanks for taking the time to look at this.

> I am able to build using the documented docker command without error.
> My `.` and `.git` directories are not owned by root.

Interesting, if you can, could you please share your `docker --version` and your OS? I would like to understand the difference with mine.

> Your user is displayed as a numeric ID, which can be caused by the user being deleted. Perhaps git is sensitive to this, hence the "dubious owner" error.

The numeric ID is only showed inside the container, because my user doesn't exist inside the golang container.

**Comment by chrisohaver** [comment]
> The numeric ID is only showed inside the container, because my user doesn't exist inside the golang container.

You are executing the  docker command from within a container?

**Comment by pabluk** [comment]
> You are executing the docker command from within a container?

no, I'm running `docker run` from a shell with my local user within a standard Ubuntu Desktop computer.

**Comment by pabluk** [comment]
Looking at [the git source code that trigger](https://github.com/git/git/blob/1fc548b2d6a3596f3e1c1f8b1930d8dbd1e30bf3/setup.c#L1484) the message "detected dubious ownership in repository at" and [its tests](https://github.com/git/git/blob/1fc548b2d6a3596f3e1c1f8b1930d8dbd1e30bf3/t/t0033-safe-directory.sh#L24) I found another workaround that doesn't require to change the ownership of the files, it only requires to set some environment variables for `safe.directory`:
```
docker run --rm -i -t -e "GIT_CONFIG_COUNT=1" -e "GIT_CONFIG_KEY_0=safe.directory" -e "GIT_CONFIG_VALUE_0=/v" -v $PWD:/v -w /v golang:1.21 make
```

**Comment by pabluk** [comment]
Just to confirm, I've created a VM with a fresh install of Ubuntu LTS 22.04 and Docker (version 24.0.5), and I was able to reproduce the same error.

<details>
<summary>Verbose console output</summary>

<pre>
root@scw-quirky-driscoll:~# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.3 LTS
Release:	22.04
Codename:	jammy
root@scw-quirky-driscoll:~# sudo -u ubuntu -i
ubuntu@scw-quirky-driscoll:~$ sudo apt update
Hit:1 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:2 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [119 kB]
Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [110 kB]
Get:4 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [109 kB]                                                    
Hit:5 https://ppa.launchpadcontent.net/scaleway/stable/ubuntu jammy InRelease              
Get:6 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [761 kB]
Get:7 http://security.ubuntu.com/ubuntu jammy-security/main Translation-en [164 kB]
Get:8 http://security.ubuntu.com/ubuntu jammy-security/main amd64 c-n-f Metadata [11,2 kB]
Get:9 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [822 kB]
Get:10 http://security.ubuntu.com/ubuntu jammy-security/restricted Translation-en [132 kB]
Get:11 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [781 kB]
Get:12 http://security.ubuntu.com/ubuntu jammy-security/universe Translation-en [143 kB]
Get:13 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 c-n-f Metadata [16,7 kB]
Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [970 kB]
Get:15 http://archive.ubuntu.com/ubuntu jammy-updates/main Translation-en [222 kB]
Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 c-n-f Metadata [15,6 kB]
Get:17 http://archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [838 kB]
Get:18 http://archive.ubuntu.com/ubuntu jammy-updates/restricted Translation-en [135 kB]
Get:19 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [979 kB]
Get:20 http://archive.ubuntu.com/ubuntu jammy-updates/universe Translation-en [213 kB]
Get:21 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 c-n-f Metadata [21,8 kB]
Get:22 http://archive.ubuntu.com/ubuntu jammy-updates/multiverse amd64 Packages [41,6 kB]
Get:23 http://archive.ubuntu.com/ubuntu jammy-backports/main amd64 Packages [41,7 kB]
Get:24 http://archive.ubuntu.com/ubuntu jammy-backports/main Translation-en [10,5 kB]
Get:25 http://archive.ubuntu.com/ubuntu jammy-backports/universe amd64 Packages [24,3 kB]
Get:26 http://archive.ubuntu.com/ubuntu jammy-backports/universe Translation-en [16,4 kB]
Get:27 http://archive.ubuntu.com/ubuntu jammy-backports/universe amd64 c-n-f Metadata [640 B]
Fetched 6 699 kB in 2s (3 866 kB/s)                                
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
23 packages can be upgraded. Run 'apt list --upgradable' to see them.
ubuntu@scw-quirky-driscoll:~$ sudo apt install docker.io
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  bridge-utils containerd dns-root-data dnsmasq-base pigz runc ubuntu-fan
Suggested packages:
  ifupdown aufs-tools cgroupfs-mount | cgroup-lite debootstrap docker-doc rinse zfs-fuse | zfsutils
The following NEW packages will be installed:
  bridge-utils containerd dns-root-data dnsmasq-base docker.io pigz runc ubuntu-fan
0 upgraded, 8 newly installed, 0 to remove and 23 not upgraded.
Need to get 69,7 MB of archives.
After this operation, 267 MB of additional disk space will be used.
Do you want to continue? [Y/n] y
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 pigz amd64 2.6-1 [63,6 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/main amd64 bridge-utils amd64 1.7-1ubuntu3 [34,4 kB]
Get:3 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 runc amd64 1.1.7-0ubuntu1~22.04.1 [4 249 kB]
Get:4 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 containerd amd64 1.7.2-0ubuntu1~22.04.1 [36,0 MB]
Get:5 http://archive.ubuntu.com/ubuntu jammy/main amd64 dns-root-data all 2021011101 [5 256 B]
Get:6 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 dnsmasq-base amd64 2.86-1.1ubuntu0.3 [354 kB]
Get:7 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 docker.io amd64 24.0.5-0ubuntu1~22.04.1 [28,9 MB]
Get:8 http://archive.ubuntu.com/ubuntu jammy/universe amd64 ubuntu-fan all 0.12.16 [35,2 kB]
Fetched 69,7 MB in 3s (24,2 MB/s)
Preconfiguring packages ...
Selecting previously unselected package pigz.
(Reading database ... 71278 files and directories currently installed.)
Preparing to unpack .../0-pigz_2.6-1_amd64.deb ...
Unpacking pigz (2.6-1) ...
Selecting previously unselected package bridge-utils.
Preparing to unpack .../1-bridge-utils_1.7-1ubuntu3_amd64.deb ...
Unpacking bridge-utils (1.7-1ubuntu3) ...
Selecting previously unselected package runc.
Preparing to unpack .../2-runc_1.1.7-0ubuntu1~22.04.1_amd64.deb ...
Unpacking runc (1.1.7-0ubuntu1~22.04.1) ...
Selecting previously unselected package containerd.
Preparing to unpack .../3-containerd_1.7.2-0ubuntu1~22.04.1_amd64.deb ...
Unpacking containerd (1.7.2-0ubuntu1~22.04.1) ...
Selecting previously unselected package dns-root-data.
Preparing to unpack .../4-dns-root-data_2021011101_all.deb ...
Unpacking dns-root-data (2021011101) ...
Selecting previously unselected package dnsmasq-base.
Preparing to unpack .../5-dnsmasq-base_2.86-1.1ubuntu0.3_amd64.deb ...
Unpacking dnsmasq-base (2.86-1.1ubuntu0.3) ...
Selecting previously unselected package docker.io.
Preparing to unpack .../6-docker.io_24.0.5-0ubuntu1~22.04.1_amd64.deb ...
Unpacking docker.io (24.0.5-0ubuntu1~22.04.1) ...
Selecting previously unselected package ubuntu-fan.
Preparing to unpack .../7-ubuntu-fan_0.12.16_all.deb ...
Unpacking ubuntu-fan (0.12.16) ...
Setting up dnsmasq-base (2.86-1.1ubuntu0.3) ...
Setting up runc (1.1.7-0ubuntu1~22.04.1) ...
Setting up dns-root-data (2021011101) ...
Setting up bridge-utils (1.7-1ubuntu3) ...
Setting up pigz (2.6-1) ...
Setting up containerd (1.7.2-0ubuntu1~22.04.1) ...
Created symlink /etc/systemd/system/multi-user.target.wants/containerd.service → /lib/systemd/system/containerd.service.
Setting up ubuntu-fan (0.12.16) ...
Created symlink /etc/systemd/system/multi-user.target.wants/ubuntu-fan.service → /lib/systemd/system/ubuntu-fan.service.
Setting up docker.io (24.0.5-0ubuntu1~22.04.1) ...
Adding group `docker' (GID 121) ...
Done.
Created symlink /etc/systemd/system/multi-user.target.wants/docker.service → /lib/systemd/system/docker.service.
Created symlink /etc/systemd/system/sockets.target.wants/docker.socket → /lib/systemd/system/docker.socket.
Processing triggers for dbus (1.12.20-2ubuntu4.1) ...
Processing triggers for man-db (2.10.2-1) ...
Scanning processes...                                                                                                                                                                                                
Scanning linux images...                                                                                                                                                                                             

Running kernel seems to be up-to-date.

No services need to be restarted.

No containers need to be restarted.

No user sessions are running outdated binaries.

No VM guests are running outdated hypervisor (qemu) binaries on this host.
ubuntu@scw-quirky-driscoll:~/coredns$ exit
logout
root@scw-quirky-driscoll:~# usermod -a -G docker ubuntu
root@scw-quirky-driscoll:~# sudo -u ubuntu -i
ubuntu@scw-quirky-driscoll:~$ id
uid=1000(ubuntu) gid=1000(ubuntu) groups=1000(ubuntu),4(adm),20(dialout),24(cdrom),25(floppy),27(sudo),29(audio),30(dip),44(video),46(plugdev),119(netdev),120(lxd),121(docker)
ubuntu@scw-quirky-driscoll:~$ git clone https://github.com/coredns/coredns.git
Cloning into 'coredns'...
remote: Enumerating objects: 64918, done.
remote: Counting objects: 100% (269/269), done.
remote: Compressing objects: 100% (177/177), done.
remote: Total 64918 (delta 134), reused 195 (delta 91), pack-reused 64649
Receiving objects: 100% (64918/64918), 99.37 MiB | 13.10 MiB/s, done.
Resolving deltas: 100% (36928/36928), done.
ubuntu@scw-quirky-driscoll:~$ cd coredns
ubuntu@scw-quirky-driscoll:~/coredns$ docker run --rm -i -t -v $PWD:/v -w /v golang:1.21 make
fatal: detected dubious ownership in repository at '/v'
To add an exception for this directory, call:

	git config --global --add safe.directory /v
CGO_ENABLED=0  go build -v -ldflags="-s -w -X github.com/coredns/coredns/coremain.GitCommit=" -o coredns
go: downloading github.com/coredns/caddy v1.1.1
go: downloading github.com/infobloxopen/go-trees v0.0.0-20200715205103-96a057b8dfb9
go: downloading github.com/miekg/dns v1.1.55
go: downloading github.com/prometheus/client_golang v1.16.0
go: downloading github.com/Azure/azure-sdk-for-go v68.0.0+incompatible
go: downloading github.com/Azure/go-autorest/autorest v0.11.29
go: downloading github.com/Azure/go-autorest/autorest/azure/auth v0.5.12
go: downloading github.com/Azure/go-autorest v14.2.0+incompatible
go: downloading k8s.io/utils v0.0.0-20230406110748-d93618cff8a2
go: downloading google.golang.org/api v0.138.0
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/dnstap/golang-dnstap v0.4.0
go: downloading github.com/farsightsec/golang-framestream v0.3.0
go: downloading google.golang.org/protobuf v1.31.0
go: downloading golang.org/x/crypto v0.12.0
go: downloading go.etcd.io/etcd/api/v3 v3.5.9
go: downloading go.etcd.io/etcd/client/v3 v3.5.9
go: downloading github.com/opentracing/opentracing-go v1.2.0
go: downloading github.com/oschwald/geoip2-golang v1.9.0
go: downloading google.golang.org/grpc v1.57.0
go: downloading github.com/go-logr/logr v1.2.4
go: downloading k8s.io/api v0.28.1
go: downloading k8s.io/apimachinery v0.28.1
go: downloading k8s.io/client-go v0.28.1
go: downloading k8s.io/klog/v2 v2.100.1
go: downloading github.com/aws/aws-sdk-go v1.45.2
go: downloading github.com/openzipkin-contrib/zipkin-go-opentracing v0.5.0
go: downloading github.com/openzipkin/zipkin-go v0.4.2
go: downloading gopkg.in/DataDog/dd-trace-go.v1 v1.54.0
go: downloading github.com/antonmedv/expr v1.13.0
go: downloading github.com/flynn/go-shlex v0.0.0-20150515145356-3f9db97f8568
go: downloading github.com/grpc-ecosystem/grpc-opentracing v0.0.0-20180507213350-8e809c8a8645
go: downloading github.com/quic-go/quic-go v0.38.1
go: downloading golang.org/x/net v0.14.0
go: downloading golang.org/x/sys v0.12.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading github.com/cespare/xxhash/v2 v2.2.0
go: downloading github.com/prometheus/client_model v0.4.0
go: downloading github.com/prometheus/common v0.44.0
go: downloading github.com/prometheus/procfs v0.10.1
go: downloading github.com/Azure/go-autorest/autorest/adal v0.9.22
go: downloading github.com/Azure/go-autorest/autorest/azure/cli v0.4.5
go: downloading github.com/Azure/go-autorest/logger v0.2.1
go: downloading github.com/dimchansky/utfbom v1.1.1
go: downloading github.com/Azure/go-autorest/tracing v0.6.0
go: downloading golang.org/x/oauth2 v0.11.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.4
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/golang/protobuf v1.5.3
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.9
go: downloading go.uber.org/zap v1.17.0
go: downloading github.com/oschwald/maxminddb-golang v1.11.0
go: downloading google.golang.org/genproto/googleapis/rpc v0.0.0-20230807174057-1744710a1577
go: downloading google.golang.org/genproto v0.0.0-20230803162519-f966b187b2e5
go: downloading github.com/google/gofuzz v1.2.0
go: downloading sigs.k8s.io/structured-merge-diff/v4 v4.2.3
go: downloading github.com/imdario/mergo v0.3.12
go: downloading github.com/spf13/pflag v1.0.5
go: downloading golang.org/x/term v0.11.0
go: downloading github.com/opentracing-contrib/go-observer v0.0.0-20170622124052-a52f23424492
go: downloading github.com/DataDog/datadog-agent/pkg/obfuscate v0.45.0-rc.1
go: downloading github.com/DataDog/datadog-go/v5 v5.1.1
go: downloading github.com/DataDog/sketches-go v1.2.1
go: downloading github.com/tinylib/msgp v1.1.8
go: downloading golang.org/x/time v0.3.0
go: downloading golang.org/x/xerrors v0.0.0-20220907171357-04be3eba64a2
go: downloading github.com/apparentlymart/go-cidr v1.1.0
go: downloading github.com/Azure/go-autorest/autorest/to v0.2.0
go: downloading github.com/Azure/go-autorest/autorest/date v0.3.0
go: downloading github.com/golang-jwt/jwt/v4 v4.5.0
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading cloud.google.com/go/compute/metadata v0.2.3
go: downloading cloud.google.com/go/compute v1.23.0
go: downloading github.com/google/s2a-go v0.1.5
go: downloading github.com/googleapis/gax-go/v2 v2.12.0
go: downloading go.opencensus.io v0.24.0
go: downloading google.golang.org/genproto/googleapis/api v0.0.0-20230803162519-f966b187b2e5
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading go.uber.org/atomic v1.11.0
go: downloading go.uber.org/multierr v1.6.0
go: downloading gopkg.in/inf.v0 v0.9.1
go: downloading sigs.k8s.io/json v0.0.0-20221116044647-bc3834ca7abd
go: downloading github.com/json-iterator/go v1.1.12
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/google/gnostic-models v0.6.8
go: downloading github.com/google/go-cmp v0.5.9
go: downloading github.com/jmespath/go-jmespath v0.4.0
go: downloading github.com/outcaste-io/ristretto v0.2.1
go: downloading github.com/philhofer/fwd v1.1.2
go: downloading github.com/DataDog/datadog-agent/pkg/remoteconfig/state v0.48.0-devel.0.20230725154044-2549ba9058df
go: downloading golang.org/x/exp v0.0.0-20221205204356-47842c84f3db
go: downloading github.com/googleapis/enterprise-certificate-proxy v0.2.5
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading k8s.io/kube-openapi v0.0.0-20230717233707-2695361300d9
go: downloading golang.org/x/text v0.12.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading sigs.k8s.io/yaml v1.3.0
go: downloading github.com/DataDog/go-tuf v1.0.1-0.5.2
go: downloading github.com/pkg/errors v0.9.1
go: downloading google.golang.org/appengine v1.6.7
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822
go: downloading github.com/go-openapi/jsonreference v0.20.2
go: downloading github.com/go-openapi/swag v0.22.3
go: downloading github.com/dustin/go-humanize v1.0.1
go: downloading github.com/secure-systems-lab/go-securesystemslib v0.7.0
go: downloading github.com/emicklei/go-restful/v3 v3.10.2
go: downloading github.com/go-openapi/jsonpointer v0.19.6
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading github.com/josharian/intern v1.0.0
# cd /v; git status --porcelain
fatal: detected dubious ownership in repository at '/v'
To add an exception for this directory, call:

	git config --global --add safe.directory /v
error obtaining VCS status: exit status 128
	Use -buildvcs=false to disable VCS stamping.
make: *** [Makefile:16: coredns] Error 1
ubuntu@scw-quirky-driscoll:~/coredns$ ls -la
total 240
drwxrwxr-x 13 ubuntu ubuntu  4096 sept.  7 22:25 .
drwxr-x---  4 ubuntu ubuntu  4096 sept.  7 22:26 ..
drwxrwxr-x  2 ubuntu ubuntu  4096 sept.  7 22:25 .circleci
-rw-rw-r--  1 ubuntu ubuntu   128 sept.  7 22:25 .codecov.yml
-rw-rw-r--  1 ubuntu ubuntu    11 sept.  7 22:25 .dockerignore
-rw-rw-r--  1 ubuntu ubuntu   247 sept.  7 22:25 .dreck.yaml
drwxrwxr-x  8 ubuntu ubuntu  4096 sept.  7 22:25 .git
drwxrwxr-x  4 ubuntu ubuntu  4096 sept.  7 22:25 .github
-rw-rw-r--  1 ubuntu ubuntu   125 sept.  7 22:25 .gitignore
-rw-rw-r--  1 ubuntu ubuntu   180 sept.  7 22:25 .golangci.yml
-rw-rw-r--  1 ubuntu ubuntu   112 sept.  7 22:25 .stickler.yml
-rw-rw-r--  1 ubuntu ubuntu   266 sept.  7 22:25 .yamllint
-rw-rw-r--  1 ubuntu ubuntu  4205 sept.  7 22:25 ADOPTERS.md
-rw-rw-r--  1 ubuntu ubuntu  2971 sept.  7 22:25 CODEOWNERS
lrwxrwxrwx  1 ubuntu ubuntu    26 sept.  7 22:25 CODE_OF_CONDUCT.md -> .github/CODE_OF_CONDUCT.md
lrwxrwxrwx  1 ubuntu ubuntu    23 sept.  7 22:25 CONTRIBUTING.md -> .github/CONTRIBUTING.md
-rw-rw-r--  1 ubuntu ubuntu   732 sept.  7 22:25 Dockerfile
-rw-rw-r--  1 ubuntu ubuntu  7868 sept.  7 22:25 GOVERNANCE.md
-rw-rw-r--  1 ubuntu ubuntu 11371 sept.  7 22:25 LICENSE
-rw-rw-r--  1 ubuntu ubuntu   742 sept.  7 22:25 Makefile
-rw-rw-r--  1 ubuntu ubuntu  2195 sept.  7 22:25 Makefile.doc
-rw-rw-r--  1 ubuntu ubuntu  4745 sept.  7 22:25 Makefile.docker
-rw-rw-r--  1 ubuntu ubuntu  6037 sept.  7 22:25 Makefile.release
-rw-rw-r--  1 ubuntu ubuntu 10738 sept.  7 22:25 README.md
lrwxrwxrwx  1 ubuntu ubuntu    19 sept.  7 22:25 SECURITY.md -> .github/SECURITY.md
drwxrwxr-x  4 ubuntu ubuntu  4096 sept.  7 22:25 core
-rw-rw-r--  1 ubuntu ubuntu  1406 sept.  7 22:25 coredns.1.md
-rw-rw-r--  1 ubuntu ubuntu   246 sept.  7 22:25 coredns.go
-rw-rw-r--  1 ubuntu ubuntu  3370 sept.  7 22:25 corefile.5.md
drwxrwxr-x  2 ubuntu ubuntu  4096 sept.  7 22:25 coremain
-rw-rw-r--  1 ubuntu ubuntu  2398 sept.  7 22:25 directives_generate.go
-rw-rw-r--  1 ubuntu ubuntu  6603 sept.  7 22:25 go.mod
-rw-rw-r--  1 ubuntu ubuntu 47953 sept.  7 22:25 go.sum
drwxrwxr-x  2 ubuntu ubuntu  4096 sept.  7 22:25 man
drwxrwxr-x  2 ubuntu ubuntu  4096 sept.  7 22:25 notes
-rw-rw-r--  1 ubuntu ubuntu  1627 sept.  7 22:25 owners_generate.go
drwxrwxr-x  2 ubuntu ubuntu  4096 sept.  7 22:25 pb
drwxrwxr-x 57 ubuntu ubuntu  4096 sept.  7 22:25 plugin
-rw-rw-r--  1 ubuntu ubuntu  1344 sept.  7 22:25 plugin.cfg
-rw-rw-r--  1 ubuntu ubuntu  6786 sept.  7 22:25 plugin.md
drwxrwxr-x  2 ubuntu ubuntu  4096 sept.  7 22:25 request
drwxrwxr-x  2 ubuntu ubuntu  4096 sept.  7 22:25 test
ubuntu@scw-quirky-driscoll:~/coredns$ docker --version
Docker version 24.0.5, build 24.0.5-0ubuntu1~22.04.1
ubuntu@scw-quirky-driscoll:~/coredns$ 
</pre>

</details>

**Comment by chrisohaver** [comment]
> ... please share your `docker --version` and your OS? I would like to understand the difference with mine.

Docker 24.0.5; MacOS 13.5.2

**Comment by polarathene** [comment]
That is a macOS volume driver issue AFAIK.

If you run without any bind volume mount you should have no issues?

You can create a container without starting it, then use `docker cp` IIRC to transfer some files into that container, then start the container and everything should work fine?

After that, you should have success with named data volume instead of a bind mount volume too, since I think that avoids the macOS volume driver used for bind mounts?

If that also works, then you'll want to look up the macOS issues upstream and driver docs. macOS has a few different storage drivers IIRC, and recently the default was changed and is causing problems like this I think (_I've had similar reports on a repo I maintain_). Thus bug upstream that should be resolved, not really something CoreDNS should need to workaround.

---

Just to add, for the purpose of building, to get the artifact back onto the host afterwards you should be able to use `docker cp` in the other direction too if you're not familiar with it. If this is something you plan on doing frequently just create a script file that runs the sequence of commands, or if possible change the affected driver Docker switched to until they resolve that bug.

**Comment by chrisohaver** [comment]
MacOS worked fine for me.

**Comment by polarathene** [comment]
PR raised with advice for users running into this scenario: https://github.com/coredns/coredns/pull/6319

**TL;DR:**
- Adjust host ownership to `root` for the git repo.
- Use `--user` with `-v /tmp/cache:/.cache` (_or similar location, but make sure ownership matches `--user`_)
- **Easiest:** On linux use `--user` and `--tmpfs` instead for `/.cache` mount (_no need to fix ownership_)
- Use a `Dockerfile` to build it or provide the source and your preferred way to output the artifact with correct ownership.

---

> MacOS worked fine for me.

Ohhh... my bad I mixed you both up sorry about that.

I reproduced the failure @pabluk experienced and did what I should have and read the original report 😓 

I don't think it affects you on macOS for the same issue I was talking about though. I recall the user reports talking about ownership being aliased to root? 🤷‍♂️ 

You've got a layer of indirection there right? If you mount anything from macOS that is your user, what does that appear as in the container?

---

## Awkward `docker cp`

If @pabluk doesn't want to change ownership to root on the host, it could be done via the container. To prevent that applying on the host due to volume mount, instead this approach can be taken:

```bash
# If not using docker run, you can create container and copy data over before starting it:
docker create --name test -w /v golang:1.21 bash -c 'chown -R root:root /v && make'

# Copy over the git repo:
docker cp ./ test:/v

# Run container with settings it was created with:
docker start -i test

# Binary built, copy it back over somewhere:
docker cp test:/v/coredns /tmp/coredns

# Test is working:
/tmp/coredns -version
```

A better approach in this situation might just be something like [`Dockerfile.builder`](https://stackoverflow.com/a/63995752), or a build stage (_that CI could either use too, or skip when building it's image_). That should avoid the issue, the build could still be performed by running the container or while building the image, output ownership won't match the non-root user, so `docker cp` or `chown` is still needed (_the image could automate the chown for `docker run --user`_).

---

> I've also tried to change the user running inside the container using `-u=$UID` but the build cache directory is by default under `/.cache` owned by `root` so it fail too

Actually... you could throw in an anonymous / named volume, or bind volume to address that too? (_**EDIT:** Nope, even with the `:nocopy` option this is mounting / creating the volume directory as `root:root`_)

**Alternatives:**
- Bind mount another local folder (_that you created with your user/group ownership for `--user`_) and that should work too, additionally persisting the cache instead of discarding it like `--tmpfs`.
- If using a `Dockerfile.build` approach, you can use the `RUN --mount=cache` for `/.cache` dir, and [`RUN --mount=bind`](https://docs.docker.com/engine/reference/builder/#run---mounttypebind) which technically allows the build to spit out an artifact without creating a container to do the same (or `docker cp`).

## `--user` + `--tmpfs`

For Linux, you can use tmpfs driver instead:

```bash
docker run --rm -it \
  --user "$(id -u):$(id -g)" \
  --tmpfs "/.cache/:uid=$(id -u),gid=$(id -g)" \
  --volume "${PWD}:/src" --workdir /src \
  golang:1.21 make
```

---

## Update

As already raised earlier in the discussion before me, you can workaround via [`git` ENV support to temporarily configure a bypass](https://github.com/coredns/coredns/issues/6303#issuecomment-1710822799):

```bash
docker run --rm -it \
  -e "GIT_CONFIG_COUNT=1" \
  -e "GIT_CONFIG_KEY_0=safe.directory" \
  -e "GIT_CONFIG_VALUE_0=/src" \
  --volume "${PWD}:/src" --workdir /src \
  golang:1.21 make
```

Or as the original error suggests by running the following before `make` (_if an image needs to support `git` at a location at container runtime, [handle it in the `Dockerfile` with a `RUN`](https://github.com/editorconfig-checker/editorconfig-checker/issues/268#issuecomment-1826200253)_):

```bash
git config --global --add safe.directory
```
